### PR TITLE
Merge Sort

### LINKED ISSUE #4148: [Entry] Algorithms: Adding Merge Sort
### Reason/inspiration (optional)

In the Algorithms Folder, I would like to add the following Sorting Algo:
- MergeSort

Within this doc, there should be 
- A brief explanation of the sorting algorithm (including time complexity):
- graphical/photographic representation 
- example code of sorting algo
- explanation of the sorting code 

### Entry Type (select all that apply)

New Term for Existing Concept Entry

### Code of Conduct

- [X] By submitting this issue, I agree to follow Codecademy Doc's [Code of Conduct](https://github.com/Codecademy/docs/blob/main/.github/CODE_OF_CONDUCT.md).

### For Maintainers

- [x] Labels added
- [x] Issue is assigned

### PR DESCRIPTION
**Pull request by JustinhSE**
### Description
Closes #4148
Included a conceptual explanation in the markdown file. I also included an implementation, visual image & runtime analysis. I found this implementation to be good if someone tried to memorize this sorting algorithm.

### Type of Change

- Adding a new entry

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [ ] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] Under "Development" on the right, I have linked any issues that are relevant to this PR (write "Closes #<issue number> in the "Description" above).

### CODE REVIEW

**Comment by JustinhSE** [comment]
@dakshdeepHERE  this is the new PR

**Comment by dakshdeepHERE** [comment]
@JustinhSE whenever you face any issue with Git ask us maintainers here. We will help us if the issue is way to irreparable then we will move to closing and reopening the PRs. :) I will move on to the reviews.

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (dakshdeepHERE) [inline content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:24]:
```suggestion
**Divide:**
```

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (dakshdeepHERE) [inline content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:26]:
```suggestion
  - `Base Case`: If the input array `data` has only one element, it's already sorted, so return.
  - `Split`: Divide the array into two halves, `left` and `right`, of approximately equal size.
```

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (dakshdeepHERE) [inline content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:34]:
```suggestion
**Conquer**:
  
  - [`Recursive Calls`](https://www.codecademy.com/learn/java-algorithms/modules/recursion-apcs/cheatsheet): Repeatedly call `mergeSort` on each half, `(left)` and `(right)`, to create subarrays so they are sorted independently.

**Combine**:
  
  - `Merge`: Call the `merge` function to merge the two sorted halves (each initially of length 1) `left` and `right`, and combine them back into the original array `data` in a sorted manner.
```

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (dakshdeepHERE) [inline content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:38]: give a basic description of the code snippet down below which can give a rough idea for the readers @JustinhSE

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (dakshdeepHERE) [inline content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:60]:
```suggestion
```

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (dakshdeepHERE) [inline content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:73]:
```suggestion
1. **Initialize**: 
   - Set indices `l`, `r`, and `k` to 0 to track the positions in `left`, `right`, and `data`.
2. **Compare and Merge**:
   - While both `left` and `right` have elements remaining:
     - Compare the elements at `left[l]` and `right[r]`.
     - Copy the smaller element into `data[k]` and increment the corresponding index (`l` or `r`).
     - Increment `k` to move to the next position in `data`.
3. **Copy Remaining Elements**:
   - If any elements remain in `left`, copy them directly into the remaining positions in `data`.
   - Similarly, if any elements remain in `right`, copy them into the remaining positions in `data`.
```

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (dakshdeepHERE) [inline content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:75]:
```suggestion
### Implementation Pt.2
```

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (dakshdeepHERE) [inline content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:110]:
```suggestion
### Time Complexity Breakdown
```

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (dakshdeepHERE) [inline content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:76]: again give a basic desc here for the code snippet down below @JustinhSE

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (dakshdeepHERE) [inline content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:126]:
```suggestion
1. **Divide**:
   - The `mergeSort` function recursively divides the array into halves until each subarray has only one element.
   - This splitting process has a time complexity of _O(log n)_ due to the repeated halving.

2. **Conquer**:
   - The base case (array of size 1) requires no sorting, so its complexity is _O(1)._

3. **Combine**:
   - The `merge` function merges two sorted subarrays into a single sorted array.
   - It iterates through both subarrays once, comparing elements and copying them to the final array.
   - This merging process takes _O(n)_ time, where n is the total number of elements being merged.
```

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (dakshdeepHERE) [inline content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:141]:
```suggestion
#### Benefits of Merge Sort:
- `Predictable Efficiency`: Merge sort maintains a time complexity of O(n log n) regardless of the initial arrangement of elements in the input array. This means it performs equally well in best-case, average-case, and worst-case scenarios.
- `Reliable for Diverse Inputs`: This consistent performance makes merge sort a dependable choice for sorting various input sizes and datasets without unexpected slowdowns due to data arrangement.
- `Efficiency for Large Datasets`: Its O(n log n) time complexity places it among the most efficient sorting algorithms, especially for large datasets, outperforming algorithms like bubble sort and [selection sort](https://www.codecademy.com/resources/docs/general/algorithm/selection-sort) that exhibit worse-case complexities of O(n^2).
```

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (dakshdeepHERE) [inline content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:15]:
```suggestion
---
Title: 'Merge Sort Algorithm' 
Description: 'Returns a sorted array using divide and conquer.' 
Subjects: 
  - 'Computer Science'
  - 'Interview Prep'
  - 'Code Foundations'
Tags:
  - 'Sorting Algorithms'
  - 'Algorithms'
CatalogContent:
  - 'learn-java'
  - 'paths/computer-science'
```

**Review by dakshdeepHERE — CHANGES_REQUESTED** [review]
I have requested few changes for you @JustinhSE to work on.

**Comment by dakshdeepHERE** [comment]
Also I would like to point out something the image that you have added is just amazing really well done @JustinhSE 🚀

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (JustinhSE) [reply content/general/concepts/algorithm/terms/merge-sort/merge-sort.md]: Just did! @dakshdeepHERE

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (JustinhSE) [reply content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:73]: okay, just added it @dakshdeepHERE

**Comment by JustinhSE** [comment]
@dakshdeepHERE Thanks so much! I enjoyed doing this entry as it is one of my favorite sorting algos! 💻 

Just finished all commits of revision

**Review by Helfese — APPROVED** [review]

**Review by dakshdeepHERE — APPROVED** [review]
LGTM 🚀

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (SSwiniarski) [inline content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:131]:
```suggestion
- The recursive calls to `mergeSort` create a log n-level tree. 
```
Generally don't want links out to competing sites.

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (SSwiniarski) [inline content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:132]:
```suggestion
  - The number of levels in this tree directly relates to how many times the array by can be divided by 2 before reaching single-element subarrays. This is equivalent to the logarithm of the array's size (log n). For example, an array of 8 elements would have 3 levels (log2 (8) = 3). _Reference the picture below for a visual understanding._
```

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (SSwiniarski) [inline content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:19]:
```suggestion

```
Only one line break

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (SSwiniarski) [inline content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:35]:
```suggestion

```
Only one line break

**Review by SSwiniarski — CHANGES_REQUESTED** [review]
@JustinhSE I did a second review and had a few comments.

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (JustinhSE) [reply content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:131]: Okay, I only added it so users could understand what a log n-level tree is because I only learned it through data structures.

File: content/general/concepts/algorithm/terms/merge-sort/merge-sort.md
Comment (JustinhSE) [reply content/general/concepts/algorithm/terms/merge-sort/merge-sort.md:132]: @SSwiniarski alittle typo, you would just need to remove `by` after "the array"

**Pull request merged**
